File: starter-packs/python/core/map_state.py
import json
import struct
from dataclasses import dataclass, field
from typing import List, Tuple
from enum import IntEnum


class ColliderType(IntEnum):
    Wall = 0
    Projectile = 1

@dataclass
class Point:
    x: float = 0.0
    y: float = 0.0
    
    def __str__(self):
        return json.dumps(self.__dict__)

@dataclass
class Collider:
    collider_type: ColliderType = ColliderType.Projectile
    positions: List[Point] = field(default_factory=list)

    def __str__(self):
        return json.dumps({
            'collider_type': self.collider_type,
            'positions': [json.loads(str(position)) for position in self.positions]
        })

@dataclass
class MapState:
    size: int                       = 0
    discrete_grid: List[List[int]]  = field(default_factory=list)
    walls: List[Collider]           = field(default_factory=list)
    save: bytearray                 = field(default_factory=bytearray)

    def __str__(self) -> str:
        return json.dumps({
            'size': self.size,
            'discrete_grid': self.discrete_grid,
            'walls': [json.loads(str(wall)) for wall in self.walls],
            'save': ' '.join([f'0x{byte:02x}' for byte in self.save])
        }, indent=4)

File: starter-packs/javascript/core/map_state.js
const ColliderType = Object.freeze({
  Wall: 0,
  Projectile: 1,
  toString(value) {
    return Object.keys(this).find((key) => this[key] ===value) || '';
  },
});

class Point {
  constructor(x = 0.0, y = 0.0) {
    this.x = x;
    this.y = y;
  }

  toString() { return `{"x":${this.x},"y":${this.y}}`; }
};


class Collider {
  constructor(colliderType = ColliderType.Projectile, positions = []) {
    this.collider_type = colliderType;
    this.positions = positions;
  }

  toString() {
    const pos_str = this.positions.map((pos) => pos.toString()).join(",");
    return `{"collider_type":${this.collider_type},"positions":[${pos_str}]}`;
  }
};

class MapState {
  constructor(size = 0, discrete_grid = null, walls = [], save = null) {
    this.size = size;
    this.discrete_grid = discrete_grid || [];
    this.walls = walls;
    this.save = save || new Uint8Array(0);
  }

  toString() {
    const walls_str = this.walls.map((wall) => wall.toString()).join(",");
    const save_str = Array.from(this.save)
      .map((byte) => `0x${byte.toString(16).padStart(2, '0')}`)
      .join(' ');

    return `{"size": ${this.size}, "discrete_grid": ${JSON.stringify(this.discrete_grid)}, "walls": [${walls_str}], "save": ${save_str}}`;
  }
};

export { ColliderType, Point, Collider, MapState };
